Port Eden's Spawn Slot implementation

zach2good opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [not yet] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Eden has a series of PRs implementing shared spawn slots that I've been promising to port. I've gone through most of them (and Eden's current live files) to start getting things into LSB. They have a lot of systems that aren't compatible with upstream that need to be jiggled around to get to work.
- TODO

## Steps to test these changes

- TODO

## TODO

- Port all the slot data Eden has so far, it's all in yaml so I'll need to write a little script to poop it into SQL where we want it (with comments, etc.)

TeoTwawki commented on sql/mob_spawn_slots.sql:
```diff
@@ -0,0 +1,16 @@
+DROP TABLE IF EXISTS `mob_spawn_slots`;
+CREATE TABLE `mob_spawn_slots` (
+  `id` int(10) unsigned NOT NULL,
+  `mobid` int(10) unsigned NOT NULL,
```
reminder: per prior discussions in discord, we really need to tie it all up without relying on mobid being in 2 tables when this is done with. it'll mean having the "slot" id available somewhere else to tie them together. That way when mobIDs change we don't have to change it in multiple places.

zach2good commented:
Current status of this: We don't want to introduce more things that are keyed onto the global IDs of specific mobs, since they can (and do, rarely) shift around!

Needs to find another mechanism that is a) easy to use, and b) is resistant to shifts

TeoTwawki commented:
> Current status of this: We don't want to introduce more things that are keyed onto the global IDs of specific mobs, since they can (and do, rarely) shift around!
> 
> Needs to find another mechanism that is a) easy to use, and b) is resistant to shifts

The quick fix is to call the "id" in this table "slotID" and then also have that same slotID in one of existing mob tables to match them on. Keeping the current structure of this, you'd add slotID to the spawn points table. We don't need to be perfect this on the 1st pass, we'll be tearing up mob tables to reorganize eventually anyway and at that future time we can flesh this out to being more than random mobid selection and use true shared spawn nodes.

TeoTwawki commented:
just changes the querying a little.

Frankie-hz commented:
We did this on Horizon and was in the process of moving it to ASB when we saw LSB was going to use a different implementation. Once this gets resolved I can port all the spawn point information we did with retail captures and how we came to each spawnset section.

We added "spawnset" to the spawn point table to identify which mobs were in which spawnset.

https://github.com/AirSkyBoat/AirSkyBoat/blob/staging/modules/custom/sql/mob_zones/02_cop/02_wilderness/002_carpenters_landing.sql

Here is an example: 
![image](https://github.com/LandSandBoat/server/assets/105882754/cf081841-fce1-4909-a3ed-e6b03b021a0c)

![image](https://github.com/LandSandBoat/server/assets/105882754/d6288a21-bb9b-4a13-bda6-6c19c189ea08)

![image](https://github.com/LandSandBoat/server/assets/105882754/e1b90bc6-9108-4eff-9905-2bda18aa11d0)

TeoTwawki commented:
@Frankie-hz that looks very similar to what I was suggesting to Zack, but I called it slotID instead of spawnSet - then we'd use that same number in the other table instead of the mob ID. Basically numbering the spawns and using that as the basis for the relation between 2 tables because that won't shift around every update.

Frankie-hz commented:
Yeah is basically the exact same thing as you said. I was trying to give you guys an idea how how most spawn sets are actually structured in zones to maybe see if there was a different way that would pop up instead.

We also added a zone column because we didnt want to structure it where if we had to add another slotid to say "Zone 2" (Using carp landing example on top) we could just add slotID 7 instead of adding it to the end of the table.

Frankie-hz commented:
@zach2good Let me know if there is anything I can do to help move this along. I would love to get this into ASB and get rid of all our modules for this <3

zach2good commented:
We're still deciding what we want slots to represent:
- One slot represents a chance for a single mob to be spawned from a list (Eden)
- One slot represents a series of chances for multiple mobs to be spawned up to a certain limit - from a list (ASB)

Frankie-hz commented:
I am gonna leave this here as a "how the system works". This is what we observed during retail testing. I will try to remember everything but I will make edits if I realized I left something out. Maybe that will help make a determination on which system to use.

- If there is a group of 3 Ord IDs next to each other and only **one** is allowed to spawn -> when a mob from that group dies it can choose one of those three mobs (including itself) to spawn. At no point will there be more than 1 orc up at a time. 
- When the zone resets (verified during maintenance) not always the same mobs will be up. It seems to random what mobs are re-spawned during a reset but the number of mobs in that area (mobs per group) have a limit on how many can be up at at time. 
- Most night time mobs share spawn slots with mobs that spawn during the day. Example: When night time hits and all the mobs are alive there will be no ghouls inside the dunes. You must kill Goblins DURING night time in order to spawn ghouls.

Building on night time: Zitah for instance has hounds that spawn from courels. But not all of the hounds spawn from the same group. Some hounds are separate of that group and will always spawn when it ticks nighttime and then will continue to respawn the same ID when killed (Basically saying its not inside that group)

- Elementals in certain zones are also grouped with each other and will spawn on certain areas of the maps.
- Beastmen campfires are zone-wide pools. This means if you kill a goblin in a campfire on one side of the map, that goblin will NEVER spawn again until other campfire goblins are killed. I observed this in Meri Mountains when goblins were not spawning for 4 days straight and only started to re-spawn when other campfires (of the same TIER) were killed. This was also observed in ronfaure.

TeoTwawki commented:
> You must kill Goblins DURING night time in order to spawn ghouls.

More specifically a respawn has to happen during night so it can do the "dice roll". You can kill'em just before as long as the next spawn happens at night.

Some good example cases to prove how things work are right in Ghelsba. Including cases where ONE of the shared spawns is the PH to an NM - which can make it seem like that Nm has a much higher cooldown or much lower pop rate than it actually does if the specific mob that is the PH doesn't get chosen (the NM isn't tied to the whole group, just that one mob!)

tl;dr Teo is corroborating what Frank just said.

InoUno committed:
Port Eden's spawn slot impl

zach2good committed:
Add workaround for incompatibilities between Clang and libstdc++

See <actions/runner-images#8659> for
more information. Currently, Clang and libstdc++ provided by the
GitHub Actions runner image are incompatible. The basic idea of
this workaround is to remove the incompatible g++ 13 and its C++
standard library and replace that with a compatible version of
libstdc++.

zach2good committed:
git rm submodule/wiki

zach2good committed:
Remove maxSpawns from spawn slots

zach2good committed:
Add blank data for spawnslotid to mob_spawn_points

zach2good committed:
Handle default-zero spawn slots

zach2good committed:
Add example spawn slots for some Orcs in Carpenter's Landing

zach2good commented:
db changes added with:
```sql
ALTER TABLE xidb.mob_spawn_points
ADD COLUMN `spawnslotid` int(10) unsigned NOT NULL DEFAULT '0' AFTER `groupid`;
```

and dbtool -> tasks -> dump table -> mob_spawn_points

zach2good commented:
SQL query to populate testing groups (so we don't have to edit anything by hand in this PR):
```sql

UPDATE xidb.mob_spawn_points SET spawnslotid = 1 WHERE mobid = 16785440;
UPDATE xidb.mob_spawn_points SET spawnslotid = 1 WHERE mobid = 16785441;
UPDATE xidb.mob_spawn_points SET spawnslotid = 1 WHERE mobid = 16785442;

UPDATE xidb.mob_spawn_points SET spawnslotid = 2 WHERE mobid = 16785446;
UPDATE xidb.mob_spawn_points SET spawnslotid = 2 WHERE mobid = 16785447;
UPDATE xidb.mob_spawn_points SET spawnslotid = 2 WHERE mobid = 16785448;

UPDATE xidb.mob_spawn_points SET spawnslotid = 3 WHERE mobid = 16785457;
UPDATE xidb.mob_spawn_points SET spawnslotid = 3 WHERE mobid = 16785458;
UPDATE xidb.mob_spawn_points SET spawnslotid = 3 WHERE mobid = 16785459;

```

zach2good commented:
I guess the only point of contention is: `AFTER `groupid`;`, where should we put the new column? I've put it after the grouped, open to other ideas. The mobid, name and packetname line up nicely with whats in the dats, so I wanted to leave them laid out as-is

TeoTwawki commented:
> I guess the only point of contention is: `AFTER `groupid`;`, where should we put the new column? I've put it after the grouped, open to other ideas. The mobid, name and packetname line up nicely with whats in the dats, so I wanted to leave them laid out as-is

I'd put it right at the end of the pos data, after `rot`. It's also fine right before pos data which is where you look to have it right now.

Frankie-hz commented:
With this implementation does every mob need a group ID and all the % chances need to add up to 100? I am following how this chance thing exactly works.

zach2good commented:
I have no idea how it works, i just pulled it in as-is, here's the relevant code:
```cpp
bool SpawnSlot::TrySpawn()
{
    // Determine which of the mobs in the group can be spawned, and if there's one spawned already
    std::vector<std::tuple<uint32, CMobEntity*>> chanceSpawns;
    std::vector<CMobEntity*>                     remainingSpawns;

    CMobEntity* allowedSpawn = nullptr;

    uint32 totalChance = 0;

    for (auto&& entry : entries)
    {
        if (entry.mob->isAlive())
        {
            allowedSpawn = entry.mob;
            break;
        }

        if (!entry.mob->m_CanSpawn)
        {
            continue;
        }

        if (entry.spawnChance > 0)
        {
            totalChance += entry.spawnChance;
            chanceSpawns.push_back({ entry.spawnChance, entry.mob });
        }
        else
        {
            remainingSpawns.push_back(entry.mob);
        }
    }

    // Don't spawn if there's another mob in this slot already spawned.
    if (allowedSpawn)
    {
        return false;
    }

    // If there are no remaining spawns available, bail out
    if (remainingSpawns.empty())
    {
        return false;
    }

    // Check for chance spawns
    if (totalChance > 0)
    {
        uint32 roll = xirand::GetRandomNumber(100);

        // Check if roll is low enough number to make one of the chance mobs to spawn.
        if (roll < totalChance)
        {
            // Find the chance spawn which matches the roll
            uint32 accumulatedRoll = 0;

            bool spawned = false;
            for (auto&& entry : chanceSpawns)
            {
                auto mob = std::get<1>(entry);
                accumulatedRoll += std::get<0>(entry);
                if (!spawned && roll < accumulatedRoll)
                {
                    allowedSpawn = mob;
                    break;
                }
            }
        }
    }

    if (!allowedSpawn)
    {
        if (!remainingSpawns.empty())
        {
            // Pick a random mob from the non-chance spawns
            allowedSpawn = xirand::GetRandomElement(remainingSpawns);
        }
        else
        {
            return false;
        }
    }

    // Finally spawn the mob
    for (auto&& entry : entries)
    {
        if (entry.mob == allowedSpawn)
        {
            entry.mob->m_AllowRespawn = true;
            entry.mob->Spawn();
        }
        else
        {
            entry.mob->m_AllowRespawn = false;
        }
    }

    return allowedSpawn != nullptr;
}
```

Frankie-hz commented:
Update on this: Eden doesn't actually use the chance variable yet. In the event that there are weighted chances to spawn they added it as a precaution in case it was found this was true. As of right now we haven't confirmed that is an actual thing but other than that the implementation looks to be in-line with what we have found as well.

We added the slotid after the mobID so you can quickly tell which spawnslot things are in without having to think very much. I don't have a preference though either way.

zach2good commented:
Now that I've completely forgotten all context as to what was going on in this PR and what I was opposing: if you're following along with what Eden is doing - we're probably good to take it too. If we end up being wrong, then we _both_ have a lot of fixing to do :D

Frankie-hz commented:
> Now that I've completely forgotten all context as to what was going on in this PR and what I was opposing: if you're following along with what Eden is doing - we're probably good to take it too. If we end up being wrong, then we _both_ have a lot of fixing to do :D

I can live with that. I will get around to testing this when I get a minute to breathe. We more or less do it the same way.

Xaver-DaRed commented:
Not needed anymore.
